src/replication.c

```diff
@@ -828,10 +828,16 @@ int masterTryPartialResynchronization(client *c) {
  * Returns C_OK on success or C_ERR otherwise. */
 int startBgsaveForReplication(int mincapa, int req) {
     int retval;
-    int socket_target = server.repl_diskless_sync && (mincapa & SLAVE_CAPA_EOF);
+    int socket_target = 0;
     listIter li;
     listNode *ln;
 
+    /* We use a socket target if slave can handle the EOF marker and we're configured to do diskless syncs.
+     * Note that in case we're creating a "filtered" RDB (functions-only) we also force socket replication
+     * to avoid overwriting the snapshot RDB file with filtered data. */
+    socket_target = (server.repl_diskless_sync || (req & SLAVE_REQ_RDB_FUNCTIONS_ONLY)) && (mincapa & SLAVE_CAPA_EOF);
+    serverAssert(socket_target || !(req & SLAVE_REQ_RDB_FUNCTIONS_ONLY));
```

Comment:
maybe assert is a bad idea? log a message and disconnect the replica?

Reply:
ohh you have a check for that elsewhere.. so this assertion is ok, but maybe misleading, let's add a comment referring to the check in startBgsaveForReplication

Reply:
Added a commnent